File: frontend/src/api/paper.ts
import client from './client';

// V3a — Paper trading (模拟盘) API (BP-V3a-005 / T3.5). All endpoints under
// /api/v1/paper. The backend lands in a later wave, so page code wraps every
// call in try/catch and degrades to empty states on 404 / empty payloads.

export interface PaperFactorConfig {
  code: string;
  weight: number;
  direction?: number;
}

export interface PaperAccountConfig {
  preset?: string;
  factors?: PaperFactorConfig[];
  top_n?: number;
  freq?: 'W' | 'M' | number;
  initial_cash?: number;
  benchmark?: string;
  liquidity_top_k?: number;
  neutralize?: 'none' | 'industry' | 'industry_mcap';
}

// Summary row from the accounts list. Return-like fields (cum_ret/drawdown)
// are fractions (0.05 = 5%), consistent with the backtest metrics elsewhere.
export interface PaperAccountBrief {
  id: number;
  name: string;
  status: string; // running / paused / stopped
  config?: PaperAccountConfig;
  latest_nav?: number | null;
  cum_ret?: number | null;
  drawdown?: number | null;
  cash?: number | null;
  open_alerts?: number | null;
}

export interface PaperPosition {
  stock_code: string;
  shares: number;
  close_price: number | null;
  market_value: number | null;
  weight: number | null;
  pnl_pct?: number | null;
}

export interface PaperAlert {
  metric: string;
  status: string; // pass / warn / fail
  value: number | null;
  check_date: string | null;
}

export interface PaperAccountDetail {
  account: PaperAccountBrief;
  positions: PaperPosition[];
  alerts: PaperAlert[];
}

export interface PaperNavPoint {
  trade_date: string;
  nav: number | null;
  daily_ret?: number | null;
  benchmark_nav?: number | null;
  drawdown?: number | null;
}

export interface PaperNavResponse {
  series: PaperNavPoint[];
  attribution: {
    excess?: number | null;
    cost_drag?: number | null;
    cash_drag?: number | null;
    selection?: number | null;
  } | null;
}

export interface PaperTrade {
  price: number | null;
  amount: number | null;
  commission: number | null;
  stamp_duty: number | null;
  transfer_fee: number | null;
}

export interface PaperOrder {
  signal_date: string | null;
  exec_date: string | null;
  stock_code: string;
  side: string; // buy / sell
  shares?: number | null;
  status: string; // filled / pending / skipped / deferred / cancelled
  fail_reason?: string | null;
  trade?: PaperTrade | null;
}

export interface PaperDriftResponse {
  weekly: {
    week: string;
    paper_ret: number | null;
    backtest_ret: number | null;
    diff: number | null;
  }[];
  cum_paper: number | null;
  cum_backtest: number | null;
  max_abs_weekly: number | null;
  holding_jaccard?: { date: string; value: number | null }[];
}

// List accounts (summary metrics included).
export const listPaperAccounts = () =>
  client
    .get<{ accounts: PaperAccountBrief[] }>('/paper/accounts')
    .then((r) => r.data);

// Create an account; returns the new id.
export const createPaperAccount = (body: {
  name: string;
  config: PaperAccountConfig;
}) =>
  client.post<{ id: number }>('/paper/accounts', body).then((r) => r.data);

// Account detail: config + current positions + unresolved alerts.
export const getPaperAccount = (id: number) =>
  client.get<PaperAccountDetail>(`/paper/accounts/${id}`).then((r) => r.data);

// Lifecycle: pause stops signal generation; stop is terminal.
export const setPaperAccountStatus = (
  id: number,
  action: 'pause' | 'resume' | 'stop',
) =>
  client
    .post<{ ok: boolean }>(`/paper/accounts/${id}/status`, { action })
    .then((r) => r.data);

// NAV series vs benchmark + return attribution over [start, end].
export const getPaperNav = (id: number, start?: string, end?: string) =>
  client
    .get<PaperNavResponse>(`/paper/accounts/${id}/nav`, {
      params: { start, end },
    })
    .then((r) => r.data);

// Rebalance order history with simulated trade details (fees).
export const getPaperOrders = (id: number, start?: string, end?: string) =>
  client
    .get<{ orders: PaperOrder[] }>(`/paper/accounts/${id}/orders`, {
      params: { start, end },
    })
    .then((r) => r.data);

// Paper vs backtest weekly-return drift over [start, end].
export const getPaperDrift = (id: number, start?: string, end?: string) =>
  client
    .get<PaperDriftResponse>(`/paper/accounts/${id}/drift`, {
      params: { start, end },
    })
    .then((r) => r.data);

// Manual catch-up run (paper_tick) for one account; omit dates to advance to
// the latest trade date.
export const tickPaperAccount = (id: number, dates?: string[]) =>
  client
    .post<{ run_id?: number; async?: boolean }>(`/paper/accounts/${id}/tick`, {
      dates,
    })
    .then((r) => r.data);

// Admin: 推进所有 running 账户的每日 paper_tick。与 factor-health run 同模式：
// 返回 {run_id, async: true}，调用方轮询 api/admin 的 getRun(runId) 直到结束。
// 后端后续波次按同模式挂载 POST /admin/paper/run。
export const runPaperTickAll = () =>
  client
    .post<{ run_id: number; async: boolean }>('/admin/paper/run')
    .then((r) => r.data);

File: backend/app/api/paper.py
"""Paper trading router (V3a / BP-V3a). Endpoints under ``/api/v1/paper``."""

from datetime import date

from fastapi import APIRouter, Body, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import get_current_user_id, get_db
from app.services import paper_service

router = APIRouter(prefix="/paper", tags=["paper"])


def _ok(data=None, msg: str = "ok") -> dict:
    from app.main import api_ok

    return api_ok(data, msg)


@router.post("/accounts")
def create_account(
    body: dict = Body(...),
    db: Session = Depends(get_db),
    _uid: int = Depends(get_current_user_id),
) -> dict:
    """Create a paper account (config defaults see paper_service.DEFAULT_CONFIG)."""
    return _ok(paper_service.create_account(db, body["name"], body.get("config")))


@router.get("/accounts")
def list_accounts(
    db: Session = Depends(get_db),
    _uid: int = Depends(get_current_user_id),
) -> dict:
    return _ok({"accounts": paper_service.list_accounts(db)})


@router.get("/accounts/{account_id}")
def get_account(
    account_id: int,
    db: Session = Depends(get_db),
    _uid: int = Depends(get_current_user_id),
) -> dict:
    return _ok(paper_service.get_account(db, account_id))


@router.post("/accounts/{account_id}/status")
def set_status(
    account_id: int,
    body: dict = Body(...),
    db: Session = Depends(get_db),
    _uid: int = Depends(get_current_user_id),
) -> dict:
    return _ok(paper_service.set_status(db, account_id, body["action"]))


@router.get("/accounts/{account_id}/nav")
def nav_history(
    account_id: int,
    start: date | None = Query(None),
    end: date = Query(...),
    db: Session = Depends(get_db),
    _uid: int = Depends(get_current_user_id),
) -> dict:
    return _ok(paper_service.nav_history(db, account_id, start, end))


@router.get("/accounts/{account_id}/orders")
def orders_history(
    account_id: int,
    start: date | None = Query(None),
    end: date | None = Query(None),
    status: str | None = Query(None),
    db: Session = Depends(get_db),
    _uid: int = Depends(get_current_user_id),
) -> dict:
    return _ok(paper_service.orders_history(db, account_id, start, end, status))


@router.get("/accounts/{account_id}/drift")
def run_drift(
    account_id: int,
    start: date = Query(...),
    end: date = Query(...),
    db: Session = Depends(get_db),
    _uid: int = Depends(get_current_user_id),
) -> dict:
    """Replay the account config as a portfolio backtest and diff NAV paths."""
    return _ok(paper_service.run_drift(db, account_id, start, end))


@router.post("/accounts/{account_id}/tick")
def tick_account(
    account_id: int,
    body: dict = Body(default=None),
    db: Session = Depends(get_db),
    _uid: int = Depends(get_current_user_id),
) -> dict:
    """Manually advance one account (catch-up replay; ≤ 20 days per call)."""
    dates = [date.fromisoformat(x) for x in (body or {}).get("dates") or []] or None
    if dates and len(dates) > 20:
        return _ok(None, msg="单次补跑最多 20 个交易日")
    results = [paper_service.paper_tick(db, account_id, d) for d in (dates or [date.today()])]
    return _ok({"results": results})
